Recover when the Flowee backend has been down.

fstromback opened the pull request:
This is the first step of implementing recovery in the Flowee backend.

It implements richer filtering abilities for the `get_block` query, which can be used to implement recovery as discussed in issue #4. In particular, we can ask Flowee: for a particular block (either by height or by hash), get all transactions touching these particular addresses. Using this, we can quite cheaply examine the (relatively few) blocks that we might have missed during some downtime.

So far, this is not complete. The flowee backend needs to actually use the new functionality as well. I plan to do this, but I think the best way of doing this would need some more data to be persisted in the database for each transaction.

There are two paths ahead here:
1. We could keep track of the current block height (it seems we do this) and persist it in the database (not sure here). So, whenever we connect to Flowee, we can see what the current height of the chain is, and examine blocks between the previously recorded height and the current height. This works fine, except when there is no previous height for whatever reason (in these situations, we probably do not have any transactions to watch either).
2. For each invoice, we could record the height of the relevant currency at the time when the invoice was created (and persist that to the database). This means that on startup we can look at all open invoices, and determine which blocks we need to examine to find out if they have been paid or not. Of course, this assumes that an invoice is not paid before it is generated (which could happen in some situations, but I think that would be treated as a proper payment anyway).

I think solution 2 is the nicest one, since it makes it very clear exactly where we need to look (even if we sometimes need to look at more blocks than in case 1), and there is more or less no risk for race conditions (e.g. exactly when do we receive the "new block" notification in relation to other messages from Flowee?). But for this to work, we need to store the block height in the transaction.

@treeman If you think this is a reasonable approach, could you help me add this functionality to the transaction/invoice structure somehow? I think it is quicker if you do that part, as you have written most of it. Then I can fix the remaining recovery code.

As a sidenote, I think it is not enough to just ask for open invoices when subscribing for addresses in the Flowee backend. We also need to subscribe to transactions in the state `:processing`, as we want to know when they are entered into a block.

treeman commented:
I'll take a look later next week. We can already do 1 quite easily, I'll have to think if I like 2 more though.

> As a sidenote, I think it is not enough to just ask for open invoices when subscribing for addresses in the Flowee backend. We also need to subscribe to transactions in the state :processing, as we want to know when they are entered into a block.

Yeah, you're right.

fstromback commented:
Yes, option 1 was my initial idea as well. I think it will work, as long as the current height is persisted in the database (I failed to figure out whether or not it was), and if there are no large splits in the chain (this could be countered by always looking back a few blocks extra). I do, however, see benefits with option 2 - then we can determine the blocks we need to examine only by looking at the set of open and processing invoices. The strong point about that is that it feels a lot easier to keep consistent. The downside is that it could make recovery slightly more expensive after a short outage (if we have some "old" invoices). Thinking about that case now is perhaps a good argument for option 1... if we have a few old invoices that were generated for some user that lost interest and no longer intend to pay them, we would have to look back for quite far with option 2, while option 1 would know the approximate length of the outage, and would only have to look at those blocks... Maybe option 1 is better in the end anyway...

treeman commented:
> Yes, option 1 was my initial idea as well. I think it will work, as long as the current height is persisted in the database (I failed to figure out whether or not it was),

It is. In `currencies.ex` for instance:

```{.elixir}
  @spec set_height!(Currency.id(), non_neg_integer) :: :ok
  def set_height!(id, height) do
    Repo.update!(Changeset.change(%Currency{id: normalize(id)}, block_height: height))
  end
```

Repo is the database.

> The downside is that it could make recovery slightly more expensive after a short outage (if we have some "old" invoices).

Yeah. While we should add a timeout to invoices, if we just go by height alone we can find transactions to them as well, while keeping work to minimum, so I do think just tracking height is the way to go.

treeman commented:
I took a look at this, and for efficiency I think we should add address filters. It will be *very* slow to filter on our side. I think we need to use these options:

```
ReuseScriptHashFilter: 40 | bool | Reuse an address filter (ScriptHash based) created on a previous GetBlock call.
SetFilterScriptHash: 41 | bytearray | Clears and add a filter address
AddFilterScriptHash: 42 | bytearray | Add a 20 byte bitcoin address.
```

I don't think we support them? (I could be wrong.)

I've pushed code that compares the current height with the old one:

```{.elixir}
  defp recover_blocks_between(prev, next) when next - prev > 1 do
    Enum.each((prev + 1)..(next - 1), fn height ->
      # FIXME do something with height here
      # Probably want to issue a "send_get_block", and parse results in an async manner.
      # We should filter by addresses from `Invoices.active_addresses(@bch)`, so we only receive
      # relevant data.
      IO.puts("TODO recover block #{height}")
    end)
  end
```

@fstromback could you try to add the filtering for the `get_blocks` command here? And add a test with fixtures for this case? As long as I have that I can take it from there if you don't want to muck around outside of the Flowee code with this.

fstromback commented:
I think I implemented them in the first commit of this PR, but I did not add any fixtures for them.

Anyway, I will proceed with the implementation. I will tell you when I have something that is either done, or something that I want you to integrate in some other part of the system.

treeman commented:
I could only find these in the commit:

```
[:outputScripts | rest] -> [{48, true} | get_output(rest)]
[:outputAddrs | rest] -> [{50, true} | get_output(rest)]
```

But if I read the docs right, they're supposed to be booleans signifying what to return.

And there is some filtering, but isn't that just for transaction filtering, while we want to filter on addresses?

fstromback commented:
I see: the "transaction filtering" in my docs is actually: "filter transactions based on the hash of the involved addresses". That's too terse on my part, I will fix that. Basically, you can give it: [{:hash, <hashed output script>}, ...]  and it will filter on addresses. The "hashed output script" is the same as the one used to subscribe to transactions.

treeman commented:
> I see: the "transaction filtering" in my docs is actually: "filter transactions based on the hash of the involved addresses". 

Oh, alright then. I'll try that out if I get there before you, but I'll prioritize some server stuff a while, so feel free to continue with it if you want.

fstromback commented:
Status update:

I implemented the logic for this (and for waiting until Flowee is ready), but I have not yet tested it. I have also added a fixture for the info message for the future test.

I intend to make a test for this and see so that it works. There are probably minor bugs still.

fstromback commented:
I have finished the tests, fixed some bugs in the implementation, and merged with master. I'm happy to merge this now!

treeman commented on lib/bitpal/backends/flowee/protocol.ex:
```diff
@@ -169,20 +169,52 @@ defmodule BitPal.Backend.Flowee.Protocol do
+  defp get_filters(filters) do
+    case filters do
+      [] -> []
+      [{:address, hash} | rest] -> [{42, %Binary{data: hash}} | get_filters(rest)]
```
This works of course, but in Elixir we typically use `Enum.map(filters, fn {:address, hash} -> ... end)` or similar

treeman reviewed (approved):
Looks good! Just a minor Elixir style thing (that you can fix if you want to, but no biggie).

I think there's a single unsigned commit it's complaining about, otherwise let's merge it.

treeman commented:
There are some conflicts with the other PR I'm working on: https://github.com/bitpal/bitpal/pull/77, but I think merging this first will be the path of least resistance.
